src/compile/data/formatparse.ts

```diff
@@ -52,77 +53,85 @@ export class ParseNode extends DataFlowNode {
+   */
+  public static makeImplicit(parent: DataFlowNode, model: Model, ancestorParse: AncestorParse) {
+    const implicit = {};
 
     if (isUnitModel(model) || isFacetModel(model)) {
       // Parse encoded fields
       model.forEachFieldDef(fieldDef => {
         if (isTimeFieldDef(fieldDef)) {
-          parse[fieldDef.field] = 'date';
+          implicit[fieldDef.field] = 'date';
         } else if (isNumberFieldDef(fieldDef)) {
-          if (calcFieldMap[fieldDef.field] || isCountingAggregateOp(fieldDef.aggregate)) {
+          if (isCountingAggregateOp(fieldDef.aggregate)) {
             return;
           }
-          parse[fieldDef.field] = 'number';
+          implicit[fieldDef.field] = 'number';
         } else if (accessPathDepth(fieldDef.field) > 1) {
           // For non-date/non-number (strings and booleans), derive a flattened field for a referenced nested field.
           // (Parsing numbers / dates already flattens numeric and temporal fields.)
-          parse[fieldDef.field] = 'flatten';
+          implicit[fieldDef.field] = 'flatten';
         }
       });
     }
 
-    // Custom parse should override inferred parse
-    const data = model.data;
-    if (data && data.format) {
-      const p = data.format.parse;
-      if (p === null) {
-        return null;  // disable parsing completely
-      } else if (p) {
-        keys(p).forEach(field => {
-          parse[field] = p[field];
-        });
+    return this.makeWithAncestors(parent, {}, implicit, ancestorParse);
+  }
+
+  /**
+   * Creates a parse node from two sets of explicit and implicit parse and updates ancestorParse.
+   */
+  public static makeWithAncestors(parent: DataFlowNode, explicit: Dict<string>, implicit: Dict<string>, ancestorParse: AncestorParse) {
+    // We should not parse what has already been parsed in a parent (explicitly or implicitly) or what has been derived (maked as "derived").
+    for (const field of keys(implicit)) {
+      const parsedAs = ancestorParse.getWithExplicit(field);
+      if (parsedAs.value !== undefined) {
+        // We always ignore derived fields even if they are implicitly defined because we expects users to create the right types.
+        if (parsedAs.explicit || parsedAs.value === implicit[field] || parsedAs.value === 'derived') {
+          delete implicit[field];
+        } else {
+          log.warn(log.message.differentParse(field, implicit[field], parsedAs.value));
+        }
       }
     }
 
-    // We should not parse what has already been parsed in a parent
-    const modelParse = model.component.data.ancestorParse;
-    keys(modelParse).forEach(field => {
-      if (parse[field] !== modelParse[field]) {
-        log.warn(log.message.differentParse(field, parse[field], modelParse[field]));
+    for (const field of keys(explicit)) {
+      const parsedAs = ancestorParse.get(field);
+      if (parsedAs !== undefined) {
+        if (parsedAs === explicit[field]) {
```

Comment:
Done

Reply:
The above part has

> // We always ignore derived fields even if they are implicitly defined because we expects users to create the right types.

The logic for this part is different, but doesn't have any comment

Reply:
Because the logic isn't special here. If a field was parsed explicitly, we don't ignore what it has been parsed as.

Reply:
There are two cases -- each of them is different.  So better comment so it's easier for reader.

Reply:
ok